### DISCUSSION TITLE
Embeddable server

### DISCUSSION THREAD

**Discussion by Kinrany** [nats-io/nats-general#85] [Ideas]
NATS server can be embedded in Go as a library.

Have you considered providing NATS as a library for other languages too? I'm particularly interested in Rust. A library is operationally much easier to start using in an existing server application.

**Comment by gcolliso** [comment]
@Jarema could you comment as our main Rust client maintainer? Thanks!

**Comment by aricart** [comment]
The NATS server is written in Go, thus if you call the correct functions, the server runs within your process. However, the other clients are just _clients_, there's no "server" role in the client.

**Reply by Kinrany** [reply]
I understand! Go code could in theory be exposed as a library to other languages via FFI, correct?

**Reply by niklasfp** [reply]
@Kinrany I'm actually kinda doing that for dotnet, I have a small go package that uses the nats server package and exposes some simple methods, that in turn is compiled to a.so and.dll utilizing go's cgo capabilities (go builmode=c-shared) and packed in a multiplatform dotnet nuget package.
So the ffi here is is just  few exported functions from the libs and used from c# in my case, I guess the same could be done for rust.

This let's me have nats-server in-process wrapped in a Nats Server class to hide away the ffi part :)
NOTE: For now I've just exposed enough to:
1. create a server with config (accepting same args as nats-server) and add a custom logger to nats in go that forwards logs via a callback
2. Start the server
3. Stop the server

This is an extract from the generated c header
```c
extern uintptr_t CreateServer(char* arguments, loggerFunc loggerFuncPtr);
extern GoInt StartServer(uintptr_t handle);
extern GoInt ShutdownServer(uintptr_t handle);
extern void FreeServer(uintptr_t handle);

```

For now I use the nats client for all management etc so still going over a socket, but loopback

This is an example on how I use the wrapper in a console test app in c#

```csharp
var services = new ServiceCollection()
    .AddLogging(c =>
        c.AddConsole())
    .AddSingleton<NatsServer.Net.NatsServer>()
    .AddSingleton(new NatsServerOptions()
        {
            Arguments = new[]
            {
                "-js",
            }
        })
        .BuildServiceProvider();

var server = services.GetRequiredService<NatsServer.Net.NatsServer>();

var cts = new CancellationTokenSource();
server.Start(cts.Token);
```

**Reply by Denrask** [reply]
Hi @niklasfp !

I'm looking to use nats-server embedded in .NET as well. As your post is over 1 year old, I'd like to know if you have more information or an example repository somewhere? :) Thanks!

**Reply by meraydin** [reply]
+1
